### ISSUE TITLE
[Bug Report] fp32 native accurate pow, exp, exp2 (was: ttnn.pow(0.999, 1) > 1 breaks Adam optimizer)

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-metal#27072]
### Describe the bug

The ```ttnn.pow()``` function incorrectly returns a value greater than 1.0 when computing ```ttnn.pow(0.999, 1)```; it returns ```ttnn.Tensor([ 1.00257], shape=Shape([1]), dtype=DataType::FLOAT32, layout=Layout::TILE)```.

This is extremely problematic for Adam optimizer since $\beta_2 = 0.999$ is the default value across all major deep learning libraries and the optimizer fails immediately in the first step when $t=1$. The math breaks down as follows:

$\hat{v}_t = \frac{v_t}{1 - \beta_2^t} < 0 \text{ when } \beta_2^t > 1$

$\theta_{t+1} = \theta_t - \alpha \cdot \frac{\hat{m}_t}{\sqrt{\hat{v}_t} + \epsilon} \rightarrow \text{undefined when } \hat{v}_t < 0$

### Steps to reproduce the issue

Run 
```
import ttnn
import torch

with ttnn.manage_device(device_id=0) as device:

    a = ttnn.from_torch(torch.tensor([1.0], dtype=torch.float32), layout = ttnn.TILE_LAYOUT, device=device)
    b = ttnn.from_torch(torch.tensor([0.999], dtype=torch.float32), layout = ttnn.TILE_LAYOUT, device=device)
    
    res = ttnn.pow(b, a)
    print(res)
```

### Expected behavior

Expected output is tensor ```ttnn.Tensor([ 0.999], shape=Shape([1]), dtype=DataType::FLOAT32, layout=Layout::TILE)``` or one close to that value - but must be less than 1.

### Please complete the following environment information

- Version of software (commit hash) https://github.com/tenstorrent/tt-metal/commit/d9e93e19e5ba85a1244eda6fc203260e360a5b63

**Comment by KalaivaniMCW** [comment]
@nmauriceTT this is related to binary pow implementation. Should we just return the input value as it is for exponent=1 cases with an additional v_if or any changes needed in rounding ?

**Comment by KalaivaniMCW** [comment]
@umalesTT 
Will `a` always be a single scalar value  or could it be multiple values that needs to be passed as a tensor ? 
Is it possible for you to pass it as a float value instead of a tensor? 

<img width="1044" height="416" alt="Image" src="https://github.com/user-attachments/assets/e93c9792-dd7d-4b70-895c-d8d663061af5" />

**Comment by umalesTT** [comment]
@KalaivaniMCW I am running Adam optimizer from Jax through Forge and in our IR's scalar is always represented as tensor - so we can't just say in this repro test that a=1, because it is not lowered that way.

**Comment by cmaryanTT** [comment]
We will look at this as part of fp32 accuracy improvements. I'd like this to be the start of a fp32 accuracy push we can do a bit of work on in the background.

@KalaivaniMCW - we should replace pow for fp32 with a version based on exp41 or exp61 from the paper we used for exp. In fp32, it's ok to have a performance impact. We should be able to get within a few fp32 ULP.

**Comment by cmaryanTT** [comment]
In this particular case, we are off by tens of thousands of fp32 ULP, so it shouldn't be difficult to get a level of precision that solves this particular case.

**Comment by fahmedchanTT** [comment]
Unscheduled for now @cmaryanTT, this might be worth picking up after Int support, unless the forge priorities shift. 

CC: @skhatriTT

**Comment by skhatriTT** [comment]
@cmaryanTT @fahmedchanTT can we please prioritize this issue? It currently does not have a workaround and is blocking end-to-end model training

C.C: @vladimirjovanovicTT

**Comment by skhatriTT** [comment]
Just checking in if there's any update

**Comment by KalaivaniMCW** [comment]
@skhatriTT  How quick do you need this ? We're still experimenting on fp32  native for exp. 
Before this, couple of other priorities have come up - like logsigmoid and Mish. 

As for the use-case mentioned in this issue, 
I tested a temporary workaround in [kalaivani/pow_1exp ](https://github.com/tenstorrent/tt-metal/compare/main...kalaivani/pow_1exp) that gives the expected result

```
if constexpr (is_fp32_dest_acc_en) {
                v_if(in1 == 1.0f) { result = in0; }
                v_endif;
 }
```

It will not affect any perf for pow-bfloat16. And as for pow-float32, the perf differs by 2% - 3% from what I tested. 

cc: @cmaryanTT  @VirdhatchaniKN

**Comment by umalesTT** [comment]
@KalaivaniMCW unfortunately this workaround won't work, because ```ttnn.pow(0.999, 2) > 1``` as well.

**Comment by fahmedchanTT** [comment]
Since this issue doesn't have a workaround on the Forge side, might need to bump up our list.

To discuss w/ @cmaryanTT  .

**Comment by KalaivaniMCW** [comment]
Got it @umalesTT !
We are testing this case with a more accurate algorithm for float32 dtype (with exp61f implementation). That would be a better solution instead of a temp fix.

**Comment by VirdhatchaniKN** [comment]
> Got it [@umalesTT](https://github.com/umalesTT) ! We are testing this case with a more accurate algorithm for float32 dtype (with exp61f implementation). That would be a better solution instead of a temp fix.

With exp61f algorithm, I get 1.0010, but expected is 0.999 or value lesser than 1

<img width="1009" height="424" alt="Image" src="https://github.com/user-attachments/assets/6b2d8023-febf-4020-9c2a-c12adaeff5e2" />

**Comment by umalesTT** [comment]
@VirdhatchaniKN yeah, we need to get less than 1. Anything above 1 no matter how small will crash the optimizer.

**Comment by KalaivaniMCW** [comment]
Since exp61f doesn't give the expected output, We'll move towards a native pow61f impl for this.

**Comment by skhatriTT** [comment]
Does the linked [PR ](https://github.com/tenstorrent/tt-metal/pull/29647)close this issue?

**Comment by umalesTT** [comment]
I tried it on this new branch and got  ```ttnn.Tensor([ 1.0026], shape=Shape([1]), dtype=DataType::FLOAT32, layout=Layout::TILE)``` (for 0.999^1 and latest commit on the branch 44377b478c2b9cad4e27066990341e7adea4b6d0)
fyi @skhatriTT

**Comment by cmaryanTT** [comment]
@umalesTT @skhatriTT - the PR cited above solved an fp32 accurate exp function, but it still proved inaccurate when used to construct an fp32 pow function. Virdhatchani is working on an fp32 native pow.

**Comment by KalaivaniMCW** [comment]
@umalesTT 
pow(a,b) = exp( b * log(a)) - this implementation better suits your test case 
the existing pow uses an optimized approach pow(a,b) = 2^(b * log2(a))

Is it possible for you to try this out and see if there are any other failure cases 
   ```
 result = ttnn.mul(x_tt, y_tt, 
        input_tensor_a_activations=[ttnn.UnaryOpType.LOG], 
        activations=[(ttnn.UnaryOpType.EXP, False)])
```

<img width="865" height="834" alt="Image" src="https://github.com/user-attachments/assets/de1cf92b-d5a7-481d-a943-3e6a0fa951bc" />

**Comment by KalaivaniMCW** [comment]
The issue with (0.999, exponent) seems to be from log and it's accuracy issue

**Comment by vladimirjovanovicTT** [comment]
@umalesTT @KalaivaniMCW I see the numbers are less than 1 so this seems ok; however the values at first glance seem significantly off from golden - can we do sth regarding this? Potentially have a less performant flavor with a higher precision?

**Comment by nmauriceTT** [comment]
> The issue with (0.999, exponent) seems to be stemming from log and it's accuracy issue

To give some more specific details: `ttnn.pow` is internally expressed as `pow(a, b) = exp2( log2(a) / ln(2)  * b)`, where `log2(a)` is its own log2 approximation (i.e. it is different from `ttnn.log2`).

The problem here is that the approximation is good on average, but 'overshoots' 1 when x -> 1, which gives `log2(0.999) > 0` (c.f. following plots of approximation):

<img width="971" height="615" alt="Image" src="https://github.com/user-attachments/assets/236db6ae-1115-4bf6-95cc-b8e38b381ce3" />

This happens because it relies on a range reduction to [1; 2] (lets us derive trend between consecutive powers of two), as opposed to a single expensive approximation on the full range. 

To fix this, the polynomial approximation on [1; 2] must follow the following constraints (on top of being accurate):
- `log(1) = 0.0...`
- `log(2) = 1.00..`

(this would also ensure that log2 continuity)

@KalaivaniMCW @VirdhatchaniKN

**Comment by VirdhatchaniKN** [comment]
> The issue with (0.999, exponent) seems to be stemming from log and it's accuracy issue

To elaborate: I have a branch `virdhatchani/pow61f` with pow 61f approach. I tried testing them up for specific values and below are my results. 

<img width="1077" height="215" alt="Image" src="https://github.com/user-attachments/assets/23464d72-a182-4058-9e40-9f68e05235a9" />

**Comment by cmaryanTT** [comment]
@VirdhatchaniKN - sync with @mcw-vishal on generating an alternative polynomial that has the endpoint constraints, but otherwise minimizes the maximum error.

**Comment by KalaivaniMCW** [comment]
> [@umalesTT](https://github.com/umalesTT) [@KalaivaniMCW](https://github.com/KalaivaniMCW) I see the numbers are less than 1 so this seems ok; however the values at first glance seem significantly off from golden - can we do sth regarding this? Potentially have a less performant flavor with a higher precision?

Yes.. the result gets worse as the exponent increases for the (0.999 , exponent) case. 
It is because of the Log result .. We're looking at an alternate implementation for Log

**Comment by VirdhatchaniKN** [comment]
@cmaryanTT , I explored different approximation techniques to enhance the logarithm approach, and REMEZ algorithm(5th order polynomial) showed clear improvement in accuracy.
Working branch : `virdhatchani/pow61f`
<img width="699" height="182" alt="Image" src="https://github.com/user-attachments/assets/cc5c08c2-c47b-4b60-bdbc-98eb72fa43a4" />

The improved log + 61f approach shows lower absolute error compared to the existing 21f method. Tested the approach for few more values. Could see better improvement particularly for values close to 1 and higher exponents.

<img width="697" height="350" alt="Image" src="https://github.com/user-attachments/assets/c7e85757-7fb8-4877-bc69-bf82752b7d81" />

**Comment by vladimirjovanovicTT** [comment]
These nrs look great @VirdhatchaniKN , we will test it.

**Comment by umalesTT** [comment]
We tested this and it works very well.

**Comment by skhatriTT** [comment]
Re-opened the issues as the changes are not in main yet

**Comment by VirdhatchaniKN** [comment]
Update : 

- Improved POW : https://github.com/tenstorrent/tt-metal/pull/30986 (PR Review stage)
- Exp FP32 : https://github.com/tenstorrent/tt-metal/pull/29647 (Need approval)

**Comment by VirdhatchaniKN** [comment]
Update :

- Improved POW : https://github.com/tenstorrent/tt-metal/pull/30986 (PR Review)
- Exp FP32 : https://github.com/tenstorrent/tt-metal/pull/29647 - Merged

**Comment by VirdhatchaniKN** [comment]
Update :

- Improved POW : #30986
  - Facing perf failures in few models. Working on it
- Exp FP32 : #29647 - Merged

**Comment by skhatriTT** [comment]
Hey @VirdhatchaniKN, just checking in to see if there's anything pending to merge this PR #30986 to main?

**Comment by VirdhatchaniKN** [comment]
> Hey [@VirdhatchaniKN](https://github.com/VirdhatchaniKN), just checking in to see if there's anything pending to merge this PR [#30986](https://github.com/tenstorrent/tt-metal/pull/30986) to main?

The PR work is complete, but we have identified a few models that are showing performance regressions compared to the main branch. We are currently investigating and debugging those cases.

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/pull/30986
Merged to main, hence closing issue
